pkg/relocatingwebhook/image_reference_relocator.go

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2019-Present Pivotal Software, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package relocatingwebhook
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"gomodules.xyz/jsonpatch/v2"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/pivotal/image-relocation/pkg/multimap"
+)
+
+type imageReferenceRelocator struct {
+	comp    multimap.Composite
+	client  client.Client
+	decoder *admission.Decoder
+}
+
+func NewImageReferenceRelocator(comp multimap.Composite) *imageReferenceRelocator {
+	return &imageReferenceRelocator{
+		comp: comp,
+	}
+}
+
+var _ ExtendedHandler = &imageReferenceRelocator{}
+
+var (
+	podResource = metav1.GroupVersionResource{Version: "v1", Resource: "pods"}
+)
+
+func (i *imageReferenceRelocator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	// Ignore non-pod resources.
+	if req.Resource != podResource {
+		return admission.Patched(fmt.Sprintf("unexpected resource (not a pod): %v", req.Resource))
+	}
+
+	rawPod := req.Object.Raw
+	pod := corev1.Pod{}
+
+	if len(rawPod) > 0 {
+		err := i.decoder.Decode(req, &pod)
+		if err != nil {
+			return admission.Errored(http.StatusBadRequest, fmt.Errorf("decoding pod: %s", err))
+		}
+	} else {
+		err := i.client.Get(ctx, types.NamespacedName{Namespace: req.Namespace, Name: req.Name}, &pod)
```

Comment:
how can you get a resource from the API Server that hasn't been created yet? I know this isn't original code, I'm just noticing it now

Reply:
I'm not sure, but I was nervous about ripping this code out. I'll check with @cppforlife.